[E-Documents Core] - Factura-E processing using new import process

AndriusAndrulevicius opened the pull request:
This pull request does not have a related issue as it's part of the delivery for development agreed directly with @altotovi

Implements Spanish Factura-E electronic document format handler to enable reading and processing of incoming purchase invoices and credit memos.

Provides XML parsing capabilities for Factura-E documents with automatic vendor lookup using VAT registration numbers, service participants, or name/address matching.

Includes support for multi-currency transactions, tax calculations, and unit of measure conversions from international standard  #codes.

Enables viewing of imported documents through dedicated purchase document reader interface.

> [!IMPORTANT]
> This Pull request depends on   @#32

Fixes #

AndriusAndrulevicius committed:
Add PINT A-NZ E-Document Read Draft Handler and Enum Extension

AndriusAndrulevicius committed:
Adds Factura-E format support for e-document processing

Implements Spanish Factura-E electronic document format handler to enable reading and processing of incoming purchase invoices and credit memos.

Provides XML parsing capabilities for Factura-E documents with automatic vendor lookup using VAT registration numbers, service participants, or name/address matching.

Includes support for multi-currency transactions, tax calculations, and unit of measure conversions from international standard codes.

Enables viewing of imported documents through dedicated purchase document reader interface.

AndriusAndrulevicius committed:
Fix E-Document Purchase Header modification to prevent unintended changes

tinestaric commented on Apps/ES/EDocumentFormats/FacturaE/app/src/Core/EDocumentFacturaEHandler.Codeunit.al:
```diff
@@ -0,0 +1,268 @@
+                Address := XMLNode.AsXmlElement().InnerText();
+    end;
+
+    local procedure TryGetUOMCodeFromInternationalCode(TextValue: Text): Code[10]
```
nit: very vague naming with TextValue and Code...

AndriusAndrulevicius committed:
Improves Factura-E handler with documentation and code optimization

Adds comprehensive XML documentation comments to all public and internal procedures to improve code maintainability and developer experience.

Optimizes XML stream handling by eliminating intermediate variable and directly using CreateInStream method call.

Enhances unit of measure conversion logic with better variable naming and clearer control flow structure.

Updates database insert operation to use non-validated insert for better performance during line processing.

tinestaric commented on Apps/ES/EDocumentFormats/FacturaE/app/src/Core/EDocumentFacturaEHandler.Codeunit.al:
```diff
@@ -70,6 +79,13 @@ codeunit 10776 "E-Document Factura-E Handler" implements IStructuredFormatReader
+    /// <param name="XmlNamespaces">The XML namespace manager for handling XML namespaces.</param>
+    /// <param name="EDocumentPurchaseHeader">The E-Document Purchase Header record to be populated.</param>
+    /// <param name="EDocument">The E-Document record containing document information.</param>
     local procedure PopulateFacturaEPurchaseInvoiceHeader(FacturaEXML: XmlDocument; XmlNamespaces: XmlNamespaceManager; var EDocumentPurchaseHeader: Record "E-Document Purchase Header"; var EDocument: Record "E-Document")
```
Just a general note: I'm not against documenting local procedures as well, but be selective there, if it adds value or not. 
My rule of thumb is that I document public and internal procedures so someone can know what they do without going into the object, but if someone is looking at the local procedure, they're already in the object, and if the code is clear enough, no need for a comment, but if it helps, then yes. 
Mainly because local procedure are more subject to change when implementation changes, and then no one updates the comments, but for public or internally public procedures, the signature is more stable.

AndriusAndrulevicius committed:
Update EDocumentFacturaEHandler.Codeunit.al

Remove documentation from local procedures

AndriusAndrulevicius committed:
Merge branch 'dev/aan/FacturaE-v2' of https://github.com/GMatuleviciute/ALAppExtensions into dev/aan/FacturaE-v2

GMatuleviciute committed:
Delete Apps/APAC/EDocumentFormats/PINT A-NZ/app/src/Core/PINTANZReadDraftHandler.Codeunit.al

GMatuleviciute committed:
Delete Apps/APAC/EDocumentFormats/PINT A-NZ/app/src/Core/PINTANZEDocReadintoDraft.EnumExt.al

AndriusAndrulevicius committed:
merge main into dev/aan/FacturaE-v2

AndriusAndrulevicius committed:
Removes PINT A-NZ format dependency from app manifest

Cleans up the dependencies section by removing the unused E-Document Format for PINT A-NZ extension reference.

Streamlines the app configuration and eliminates unnecessary dependency declarations.

AndriusAndrulevicius committed:
Update Apps/ES/EDocumentFormats/FacturaE/app/src/Core/EDocumentFacturaEHandler.Codeunit.al

Co-authored-by: Grasiele Matuleviciute <131970463+GMatuleviciute@users.noreply.github.com>

AndriusAndrulevicius committed:
Update Apps/ES/EDocumentFormats/FacturaE/app/src/Core/EDocumentFacturaEHandler.Codeunit.al

Co-authored-by: Grasiele Matuleviciute <131970463+GMatuleviciute@users.noreply.github.com>

AndriusAndrulevicius committed:
Reorganizes imports and locks namespace labels

Reorders using statements alphabetically for better code organization.

Adds Locked = true attribute to namespace labels to prevent translation, ensuring XML namespace URIs remain constant across different language configurations.

AndriusAndrulevicius committed:
Renames label constants to use 'Tok' suffix convention

Updates variable naming to follow AL coding standards where token constants use 'Tok' suffix instead of 'Lbl' suffix.

Changes apply to namespace URI constants and XPath expression constants used in XML processing.

AndriusAndrulevicius committed:
Add Factura-E test suite and draft reset support

Introduces a mock e-document format, validation helpers, and a sample Facturae XML to enable end-to-end tests for incoming invoice parsing and purchase document creation.

Adds a reusable test library to set up services, workflows, master data, and mappings, and to drive inbound/import scenarios. Updates the test app manifest (ID range, target, resource folders) and includes a test dependency.

Adds a routine to reset drafts by deleting the generated purchase draft, enabling clean test reruns and easier state management.

AndriusAndrulevicius committed:
missing codeunit added

GMatuleviciute commented on Apps/ES/EDocumentFormats/FacturaE/test/src/EDocumentStructuredTests.Codeunit.al:
```diff
@@ -0,0 +1,268 @@
+        Assert: Codeunit Assert;
+        LibraryVariableStorage: Codeunit "Library - Variable Storage";
+        LibraryEDoc: Codeunit "Library - E-Document";
+        // EDocImplState: Codeunit "E-Doc. Impl. State";
```
```suggestion
```

GMatuleviciute commented on Apps/ES/EDocumentFormats/FacturaE/test/src/EDocumentStructuredTests.Codeunit.al:
```diff
@@ -0,0 +1,268 @@
+    var
+        EDocument: Record "E-Document";
+    begin
+        Initialize(Enum::"Service Integration"::"No Integration");
```
Feture, Scenario, Given, When, Then missing

GMatuleviciute commented on Apps/ES/EDocumentFormats/FacturaE/test/src/EDocumentStructuredTests.Codeunit.al:
```diff
@@ -0,0 +1,268 @@
+            FacturaEStructValidations.SetMockCurrencyCode(MockCurrencyCode);
+            FacturaEStructValidations.SetMockDate(MockDate);
+            FacturaEStructValidations.AssertFullEDocumentContentExtracted(EDocument."Entry No");
+        end
+        else
```
```suggestion
        end else
```

GMatuleviciute commented on Apps/ES/EDocumentFormats/FacturaE/test/src/EDocumentStructuredTests.Codeunit.al:
```diff
@@ -0,0 +1,268 @@
+        Vendor.Modify(true);
+
+        SetupFacturaEEDocumentService();
+        CreateInboundEDocumentFromXML(EDocument, 'factura-e/facturae-invoice-0.xml');
```
Same hardcoded value used a few times in the pr. Please move to global lable, lock and reuse in test procedures

GMatuleviciute commented on Apps/ES/EDocumentFormats/FacturaE/test/src/EDocumentStructuredTests.Codeunit.al:
```diff
@@ -0,0 +1,268 @@
+    begin
+        LibraryLowerPermission.SetOutsideO365Scope();
+        LibraryVariableStorage.Clear();
+        // Clear(EDocImplState);
```
```suggestion
```

GMatuleviciute commented on Apps/ES/EDocumentFormats/FacturaE/test/src/EDocumentStructuredTests.Codeunit.al:
```diff
@@ -0,0 +1,268 @@
+        LibraryEDoc.SetupStandardPurchaseScenario(Vendor, EDocumentService, Enum::"E-Document Format"::"Factura-E 3.2.2", Integration);
+        EDocumentService."Import Process" := "E-Document Import Process"::"Version 2.0";
+        EDocumentService."Read into Draft Impl." := "E-Doc. Read into Draft"::"Factura-E";
+        EDocumentService.Modify();
```
true/false

GMatuleviciute commented on Apps/ES/EDocumentFormats/FacturaE/test/src/EDocumentStructuredTests.Codeunit.al:
```diff
@@ -0,0 +1,268 @@
+        EDocLogRecord := EDocumentLog.InsertLog(Enum::"E-Document Service Status"::Imported, Enum::"Import E-Doc. Proc. Status"::Readable);
+
+        EDocument."Structured Data Entry No." := EDocLogRecord."E-Doc. Data Storage Entry No.";
+        EDocument.Modify();
```
true/false

GMatuleviciute commented on Apps/ES/EDocumentFormats/FacturaE/test/src/EDocumentStructuredTests.Codeunit.al:
```diff
@@ -0,0 +1,268 @@
+
+    local procedure SetupFacturaEEDocumentService()
+    begin
+        EDocumentService."Read into Draft Impl." := "E-Doc. Read into Draft"::"Factura-E";
+        EDocumentService.Modify();
```
true/false

GMatuleviciute commented on Apps/ES/EDocumentFormats/FacturaE/test/src/EDocumentStructuredTests.Codeunit.al:
```diff
@@ -0,0 +1,268 @@
+        if IsInitialized then
+            exit;
+
+        EDocument.DeleteAll();
+        EDocumentServiceStatus.DeleteAll();
+        EDocumentService.DeleteAll();
+        EDocDataStorage.DeleteAll();
+        EDocumentPurchaseHeader.DeleteAll();
+        EDocumentPurchaseLine.DeleteAll();
+        DocumentAttachment.DeleteAll();
```
true/false

AndriusAndrulevicius commented on Apps/ES/EDocumentFormats/FacturaE/test/src/EDocFormatMock.Codeunit.al:
```diff
@@ -0,0 +1,56 @@
+    local procedure OnGetCompleteInfoFromReceivedDocument(var EDocument: Record "E-Document"; var CreatedDocumentHeader: RecordRef; var CreatedDocumentLines: RecordRef; var TempBlob: codeunit "Temp Blob");
+    begin
+    end;
+
+
```
Copied from MSFT

AndriusAndrulevicius commented on Apps/ES/EDocumentFormats/FacturaE/test/app.json:
```diff
@@ -51,13 +51,17 @@
   },
-  "application": "27.0.0.0"
+  "application": "27.0.0.0",
+    "target": "OnPrem",
```
Used Core tests as an example.
Removed

AndriusAndrulevicius committed:
Optimize FacturaE tests; drop OnPrem target

Removes OnPrem target to align tests with SaaS environments.

Skips triggers on bulk delete/modify during setup to speed execution and reduce side effects/permission issues, improving test stability.

Centralizes the test XML path in a reusable token and adds scenario-style comments for readability, plus minor cleanup.

AndriusAndrulevicius committed:
Normalizes EOF whitespace in tests

Trims trailing blank line and removes final newline to align with repository formatting conventions.

No functional changes.
